src/core/unit.c

```diff
@@ -2607,6 +2627,29 @@ static bool unit_process_job(Job *j, UnitActiveState ns, bool reload_success) {
         return unexpected;
 }
 
+static void unit_check_concurrency_limit(Unit *u) {
+        Unit *slice = UNIT_GET_SLICE(u);
+
+        if (!slice)
+                return;
+
+        /* If a unit was stopped, maybe it has pending siblings that can be started now */
+
+        if (SLICE(slice)->concurrency_soft_max != UINT_MAX) {
+                Unit *sibling;
+                UNIT_FOREACH_DEPENDENCY(sibling, slice, UNIT_ATOM_SLICE_OF) {
+
+                        if (!sibling->job)
+                                continue;
+
+                        job_add_to_run_queue(sibling->job);
```

Comment:
I think this also needs to go down the tree in case the sibling is a slice. Suppose we have the following hierarchy:
```
a.slice -+- a-b.slice --- d.service (running)
         +- a-c.slice --- e.service (waiting)
         ...
```
`a.slice` sets a limit, the nested slices do not. If `d.service` stops, `e.service`'s job should be rescheduled, but currently it won't be, IIUC.